src/Orleans/Configuration/ProviderConfiguration.cs

```diff
@@ -193,23 +193,39 @@ public IList<IProvider> Children
     [Serializable]
     public class ProviderCategoryConfiguration
     {
-        public string Name { get; set; }
+        public ProviderCategoryConfiguration()
+        {
+            Providers = new Dictionary<string, IProviderConfiguration>();
+        }
 
-        public IDictionary<string, IProviderConfiguration> Providers { get; set; }
+        internal ProviderCategoryConfiguration(string name) : this()
```

Comment:
Changing the visibility would still leave it to be public to the other classes in the assembly.

About the noise. As you might notice, I'm more on the side "correct by construction". I see it like this: Now in this case also null is allowed and the code needs to work correctly. All the code, even the pieces that are touched by this variable and goes out of the class -- the side-effects included. If this is not the case, why not to disallow it? If it is allowed, what is the purpose? There are now two possible ways to construct this class with regard to this, in this particular case.

Suppose we want to use [FsCheck](https://fsharp.github.io/FsCheck/) or [Pex](http://research.microsoft.com/en-us/projects/pex/) (or [Smart Unit Tests](http://blogs.msdn.com/b/visualstudioalm/archive/2014/12/11/smart-unit-tests-a-mental-model.aspx)), we test this implementation and find bugs. Should they be fixed or should we either comment or assume it will not be called with invalid parameters? Absence of such frameworks, and since it takes some "brain cycles" to start to form a dependency graph, and especially since this is a constructor and by default everything in this class is affected (and everything that goes out from the class surface), it becomes somewhat more difficult to be assured this will always work. Especially in cases this will be refactored in the future. Even for provers such as Pex, it would help to trim down the dependency graph by stating which parameter ranges are allowed, I believe.

I do occasionally check invariants even in Private static fields which take parameters and return a value. Just to state the allowed range. One's mileage naturally vary, so hence the issue about contributor guidelines.